datastore/datastore.go

```diff
@@ -0,0 +1,260 @@
+	hook "github.com/alanshaw/ipfs-hookds"
+	hopts "github.com/alanshaw/ipfs-hookds/opts"
+	hres "github.com/alanshaw/ipfs-hookds/query/results"
+	hropts "github.com/alanshaw/ipfs-hookds/query/results/opts"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	levelds "github.com/ipfs/go-ds-leveldb"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+	"github.com/libp2p/go-libp2p-kad-dht/providers"
+	"github.com/libp2p/hydra-booster/metrics"
+	"github.com/multiformats/go-base32"
+	"github.com/whyrusleeping/timecache"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+// root namespace of provider keys
+var providersRoot = datastore.NewKey(providers.ProvidersKeyPrefix)
+
+// AddProviderFunc adds a provider for a given CID to the datastore
+type AddProviderFunc func(ctx context.Context, c cid.Cid, id peer.ID)
+
+// GetRoutingFunc is a function that returns an appropriate routing module given a CID
+type GetRoutingFunc func(cid.Cid) (routing.Routing, AddProviderFunc, error)
+
+// Options are options for the Hydra datastore
+type Options struct {
+	// total number of find provider queries we should queue
+	FindProvidersQueueSize int
+	// number of providers to find when a provider record does not exist in the store
+	FindProvidersCount int
+	// number of find provider requests we will concurrently process
+	FindProvidersConcurrency int
+	// maximum time a find providers call is allowed to take
+	FindProvidersTimeout time.Duration
+	// period after a find failure that another find request for the same key will be discarded
+	FindProvidersFailureBackoff time.Duration
+}
+
+// option defaults
+const (
+	findProvidersQueueSize   = 1000
+	findProvidersCount       = 1
+	findProvidersConcurrency = 1
+	findProvidersTimeout     = time.Second * 10
+)
+
+type findResult struct {
+	key      datastore.Key
+	status   string
+	duration time.Duration
+	err      error
+}
+
+// NewDatastore creates a new datastore that adds hooks to perform hydra things
+func NewDatastore(ctx context.Context, path string, getRouting GetRoutingFunc, opts Options) (datastore.Batching, error) {
+	opts = setOptionDefaults(opts)
+	ds, err := levelds.NewDatastore(path, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create datastore: %w", err)
+	}
+
+	return hook.NewBatching(ds, hopts.OnAfterQuery(newOnAfterQueryHook(ctx, getRouting, opts))), nil
+}
+
+func setOptionDefaults(opts Options) Options {
+	if opts.FindProvidersConcurrency == 0 {
+		opts.FindProvidersConcurrency = findProvidersConcurrency
+	}
+	if opts.FindProvidersCount == 0 {
+		opts.FindProvidersCount = findProvidersCount
+	}
+	if opts.FindProvidersQueueSize == 0 {
+		opts.FindProvidersQueueSize = findProvidersQueueSize
+	}
+	if opts.FindProvidersTimeout == 0 {
+		opts.FindProvidersTimeout = findProvidersTimeout
+	}
+	if opts.FindProvidersFailureBackoff == 0 {
+		opts.FindProvidersFailureBackoff = time.Minute
+	}
+	return opts
+}
```

Comment:
Could this be an anonymous func?

Reply:
Yes, I could also put it back in the constructor or we could use functional options.

I juts moved it out of the constructor because it was taking up a lot of space and making it harder to quickly scan.